runtime-transaction/src/signature_details.rs

```diff
@@ -0,0 +1,175 @@
+    solana_sdk::{message::TransactionSignatureDetails, pubkey::Pubkey},
+    solana_svm_transaction::instruction::SVMInstruction,
+};
+
+/// Get transaction signature details.
+pub fn get_signature_details<'a>(
+    num_transaction_signatures: u64,
+    instructions: impl Iterator<Item = (&'a Pubkey, SVMInstruction<'a>)>,
+) -> TransactionSignatureDetails {
+    let mut filter = SignatureDetailsFilter::new();
+
+    // Wrapping arithmetic is safe below because the maximum number of signatures
+    // per instruction is 255, and the maximum number of instructions per transaction
+    // is low enough that the sum of all signatures will not overflow a u64.
+    let mut num_secp256k1_instruction_signatures: u64 = 0;
+    let mut num_ed25519_instruction_signatures: u64 = 0;
+    for (program_id, instruction) in instructions {
+        let program_id_index = instruction.program_id_index;
+        match filter.is_signature(program_id_index, program_id) {
+            ProgramIdStatus::NotSignature => {}
+            ProgramIdStatus::Secp256k1 => {
+                num_secp256k1_instruction_signatures = num_secp256k1_instruction_signatures
+                    .wrapping_add(u64::from(get_num_signatures_in_instruction(&instruction)));
+            }
+            ProgramIdStatus::Ed25519 => {
+                num_ed25519_instruction_signatures = num_ed25519_instruction_signatures
+                    .wrapping_add(u64::from(get_num_signatures_in_instruction(&instruction)));
+            }
+        }
+    }
+
+    TransactionSignatureDetails::new(
+        num_transaction_signatures,
+        num_secp256k1_instruction_signatures,
+        num_ed25519_instruction_signatures,
+    )
+}
+
+#[inline]
+fn get_num_signatures_in_instruction(instruction: &SVMInstruction) -> u8 {
+    instruction.data.first().copied().unwrap_or(0)
+}
+
+#[derive(Copy, Clone)]
+enum ProgramIdStatus {
+    NotSignature,
+    Secp256k1,
+    Ed25519,
+}
+
+struct SignatureDetailsFilter {
+    // array of slots for all possible static and sanitized program_id_index,
+    // each slot indicates if a program_id_index has not been checked, or is
+    // already checked with result that can be reused.
+    flags: [Option<ProgramIdStatus>; FILTER_SIZE as usize],
+}
+
+impl SignatureDetailsFilter {
+    #[inline]
+    fn new() -> Self {
+        Self {
+            flags: [None; FILTER_SIZE as usize],
+        }
+    }
+
+    #[inline]
+    fn is_signature(&mut self, index: u8, program_id: &Pubkey) -> ProgramIdStatus {
+        let flag = &mut self.flags[usize::from(index)];
+        match flag {
+            Some(status) => *status,
+            None => {
+                *flag = Some(Self::check_program_id(program_id));
+                *flag.as_ref().unwrap()
+            }
+        }
+    }
+
+    #[inline]
+    fn check_program_id(program_id: &Pubkey) -> ProgramIdStatus {
+        if program_id == &solana_sdk::secp256k1_program::ID {
```

Comment:
curious why you'd choose `==` instead of `check_id()`, thought the later encapsulates some id details.

Reply:
I have a detailed explanation! - /rant triggered 😆 

`check_id` is defined via a macro as this:
```rust
        /// Returns `true` if given pubkey is the program ID.
        // TODO make this const once `derive_const` makes it out of nightly
        // and we can `derive_const(PartialEq)` on `Pubkey`.
        pub fn check_id(id: &#pubkey_type) -> bool {
            id == &ID
        }
```
because these are defined without `#[inline]` annotations, and we (typically) do not use `lto`, they **cannot** be inlined except in the compilation unit (not bigger than crate) they are defined in!

So by using `check_id` we add function call overhead ontop of our 32 byte comparisons. We could add `#[inline]` annotations to `check_id` derivation, but unclear we really want it expanded **everywhere** since it would result in a larger binary.

Based on that, where could we expect it to better than `check_id`?
We'd expect it to better when we must check many ids, since those checks are more efficient.

**But** we'd expect it to do worse in the case where we are load in this larger assembly but don't actually do many comparisons - which I think happens in the case of repeated pubkeys.

Didn't test this before, but I have those benches so I'll let them demonstrate!

<details>
<summary>Benchmark Results</summary>

```
bench_get_signature_details_empty/0 instructions
                        time:   [247.60 ps 247.79 ps 248.05 ps]
                        thrpt:  [4.0314 Gelem/s 4.0356 Gelem/s 4.0387 Gelem/s]
                 change:
                        time:   [-0.0004% +0.5805% +1.1704%] (p = 0.07 > 0.05)
                        thrpt:  [-1.1568% -0.5772% +0.0004%]
                        No change in performance detected.

bench_get_signature_details_no_sigs_unique/4 instructions
                        time:   [8.7759 ns 8.7790 ns 8.7827 ns]
                        thrpt:  [113.86 Melem/s 113.91 Melem/s 113.95 Melem/s]
                 change:
                        time:   [+31.827% +32.136% +32.432%] (p = 0.00 < 0.05)
                        thrpt:  [-24.489% -24.320% -24.143%]
                        Performance has regressed.

bench_get_signature_details_no_sigs_unique/32 instructions
                        time:   [72.539 ns 72.602 ns 72.675 ns]
                        thrpt:  [13.760 Melem/s 13.774 Melem/s 13.786 Melem/s]
                 change:
                        time:   [+77.608% +78.024% +78.451%] (p = 0.00 < 0.05)
                        thrpt:  [-43.962% -43.828% -43.696%]
                        Performance has regressed.

bench_get_signature_details_packed_sigs/4 instructions
                        time:   [5.0790 ns 5.0835 ns 5.0889 ns]
                        thrpt:  [196.51 Melem/s 196.71 Melem/s 196.89 Melem/s]
                 change:
                        time:   [-20.308% -20.052% -19.787%] (p = 0.00 < 0.05)
                        thrpt:  [+24.668% +25.082% +25.483%]
                        Performance has improved.

bench_get_signature_details_packed_sigs/64 instructions
                        time:   [27.624 ns 27.640 ns 27.659 ns]
                        thrpt:  [36.154 Melem/s 36.179 Melem/s 36.200 Melem/s]
                 change:
                        time:   [-58.254% -58.113% -57.957%] (p = 0.00 < 0.05)
                        thrpt:  [+137.85% +138.74% +139.54%]
                        Performance has improved.

bench_get_signature_details_mixed_sigs/4 instructions
                        time:   [8.2276 ns 8.2311 ns 8.2362 ns]
                        thrpt:  [121.42 Melem/s 121.49 Melem/s 121.54 Melem/s]
                 change:
                        time:   [+15.451% +15.744% +16.039%] (p = 0.00 < 0.05)
                        thrpt:  [-13.822% -13.603% -13.383%]
                        Performance has regressed.

bench_get_signature_details_mixed_sigs/64 instructions
                        time:   [38.007 ns 38.046 ns 38.098 ns]
                        thrpt:  [26.248 Melem/s 26.284 Melem/s 26.311 Melem/s]
                 change:
                        time:   [-28.501% -28.252% -28.027%] (p = 0.00 < 0.05)
                        thrpt:  [+38.941% +39.376% +39.862%]
                        Performance has improved.
```

</details>

^ Summarizing those results, basically confirms the suspicions. `check_id` does worse in cases where we do several unique pubkey comparisons, but better in cases where we do something like packed ed25519 calls.

You probably have more data that me here wrt mnb traffic - do you know how many unique programs a transaction uses on average?

Reply:
> rant triggered 

🤣 good rant. Was hoping that `check_id()` might change one day, so all call site will upgrade. Guess there isn't much to improve that macro :)

> do you know how many unique programs a transaction uses on average?

Don't know that stat handy, would guess no more than a dozen?